Delete kata-remote runtime class when peerpods gets disabled

gkurz opened the pull request:
We want peerpods to be fully turned off if enablePeerPods transitions from true to false. From a user POV this means that the kata-remote runtime class shouldn't be available anymore. 

Adapt the KataConfig install path to support this transition.

Fixes: https://issues.redhat.com/browse/KATA-4109

<!--
If this is a bug fix, make sure your description includes "Fixes: #xxxx", or
"Closes: #xxxx"

Please provide the following information:
-->
**- Description of the problem which is fixed/What is the use case**

**- What I did**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

gkurz committed:
Delete kata-remote runtime class when peerpods gets disabled

We want peerpods to be fully turned off if enablePeerPods
transitions from true to false. From a user POV this means
that the kata-remote runtime class shouldn't be available
anymore.

Delete it explicitly instead of leaving that to
the garbage collector on KataConfig deletion.

Signed-off-by: Greg Kurz <groug@kaod.org>

gkurz committed:
Move peer-pods code to its own file

$ wc -l controllers/openshift_controller.go
2595 controllers/openshift_controller.go

The file is huge already.

The peer-pods logic needs some heavy fixing that will require to add some
more peer-pods specific code.

Let's move peer-pods to a separate file. It will be easier to read and
maintain. This is consistent with what we already do for other more recent
features.

This doesn't change behavior.

Signed-off-by: Greg Kurz <groug@kaod.org>

gkurz committed:
Refactor peer pods enablement code

Consolidate the peer pods enablement code in a separate function.

This doesn't change behavior.

Signed-off-by: Greg Kurz <groug@kaod.org>

gkurz committed:
Introduce listKataPodsInRuntimeClass()

We will need to check kata pods running in a specific runtime class.

No changes in behavior.

Signed-off-by: Greg Kurz <groug@kaod.org>

gkurz committed:
Disable peer pods when KataConfig.enablePeerPods is set to false

In order to support this transition, we should check that no pods
are using the kata-remote runtime class.

Even if a pod can technically continue to run when the runtime
class is removed, this would prevent k8s to properly handle pod
restarts/rescheduling or deployments.

Signed-off-by: Greg Kurz <groug@kaod.org>

bpradipt commented on controllers/peerpods.go:
```diff
@@ -0,0 +1,482 @@
+/*
```
@gkurz nice.. Long due..

bpradipt reviewed (approved):
/lgtm

gkurz commented:
Converted back to draft as it needs some more testing on my side before we merge.

beraldoleal commented on controllers/peerpods.go:
```diff
@@ -552,6 +552,57 @@ func (r *KataConfigOpenShiftReconciler) processPeerPods(enable bool) (ctrl.Resul
 
 		// Reset the in progress condition
 		r.resetInProgressCondition()
+	} else {
```
For semantic clarity, I would recommend to split this into `enablePeerpods()` and `disablePeerpods()`.

Calling `processPeerpods(false)` doesn’t make the intent obvious without checking the signature, and separate methods would also trim down the lines here.

beraldoleal replied:
And I just noticed we have already a disablePeerpods(). So maybe move that there? Anyway, take this as nit suggestion, if isn't trivial to fix, that is fine.

gkurz replied:
I'll look into that.

beraldoleal commented on controllers/peerpods.go:
```diff
@@ -570,6 +570,17 @@ func (r *KataConfigOpenShiftReconciler) processPeerPods(enable bool) (ctrl.Resul
+			r.Log.Info("Kata Peer PODs are present. Requeue for reconciliation ")
+			return ctrl.Result{Requeue: true, RequeueAfter: 15 * time.Second}, err
+		}
+
```
Just to check: if a pod is still running, do we keep requeuing forever? That might leave us in a “blocked” state with noisy logs?

Also, since the RuntimeClass stays in place, would new pods still be scheduled?

gkurz replied:
Yes for all questions.

Rationale is that it is recommended to first delete pods and possibly reschedule them to some other runtime class before deleting the runtime class they're currently using. This requires the KataConfig owner (admin) to sync with the pod owners (users).

beraldoleal replied:
Thanks. We will just need to document this properly to let admins know this is not a kill switch.

pmores replied:
This behaviour is not new BTW.

gkurz replied:
> This behaviour is not new BTW.

Yes this behavior is already [documented](https://docs.redhat.com/fr/documentation/openshift_sandboxed_containers/1.10/html/deploying_openshift_sandboxed_containers/uninstalling-osc) for the full uninstallation of kata (AKA delete `KataConfig`). This change applies the same logic for the uninstallation of peer pods only.

beraldoleal reviewed (commented):
Hi @gkurz , thanks for this. Appreciate the refactoring part. Always good to see those.

I left a nit suggestion and some questions for you.

gkurz committed:
Refactor peer pods disablement code

Consolidate the peer pods disablement code in a separate function.

Ideally we'd want to move all the code to disablePeerPods for
symmetry with enablePeerPods. Unfortunately, this would require
to change the function signature and thus make the patch bigger.
To keep the review simple, rename the existing disablePeerPods so
that the name is available for the new code.

Signed-off-by: Greg Kurz <groug@kaod.org>

gkurz committed:
Stop using types from sigs.k8s.io/controller-runtime/pkg/reconcile

Use aliases defined in the top level sigs.k8s.io/controller-runtime
instead.

Signed-off-by: Greg Kurz <groug@kaod.org>

gkurz commented:
@beraldoleal @bpradipt PTAL again.

Notable changes :
- stop using `sigs.k8s.io/controller-runtime/pkg/reconcile` directly (cleanup)
- separate enable and disable path as asked by @beraldoleal 
- fix return path for reschedule requests

I'm leaving the PR in draft mode as I'm still having problem testing this in my setup...

pmores commented on controllers/openshift_controller.go:
```diff
@@ -333,261 +318,6 @@ func (r *KataConfigOpenShiftReconciler) removeLogLevel() error {
-	}
-}
-
-func (r *KataConfigOpenShiftReconciler) processDaemonsetForMonitor() *appsv1.DaemonSet {
```
This function is not peer pods though, is it?

gkurz replied:
Oops you're right... I fumbled my cut-n-paste obviously :wink:. Good catch @pmores !

gkurz commented:
After discussing with @snir911 , another thing to do is to make sure the credentials controllers doesn't delete the peer pods secret *before* the podvm image deletion is started. Doable with finalizers.
